typing/typeopt.ml

```diff
@@ -119,7 +119,8 @@ let classify env ty : classification =
 
 let array_type_kind env ty =
   match scrape_poly env ty with
-  | Tconstr(p, [elt_ty], _) when Path.same p Predef.path_array ->
+  | Tconstr(p, [elt_ty], _)
+    when Path.same p Predef.path_array || Path.same p Predef.path_iarray ->
```

Comment:
So immutable arrays also use dynamic unboxing of float elements? Are we sure that this is the right design choice? I wonder if people would not prefer immutable arrays without dynamic unboxing.

Reply:
Right now, immutable arrays are implemented as just a fresh interface on top of mutable arrays, and there are (free, unsafe) conversions back and forth between mutable and immutable arrays. If we removed dynamic float unboxing from immutable arrays, then we would have a way of making mutable arrays of boxed floats. I guess we could check during the conversion and error if someone tries to convert an immutable array of floats to a mutable one? Or maybe it's OK to accidentally have mutable arrays of boxed floats?

Reply:
For what it's worth, I asked around within Jane Street on this point, and (despite sometimes running into trouble with the dynamic float unboxing) we think it's best to keep separate concerns separate. Immutable arrays are simplest when they are just like mutable ones, but immutable (and covariant). Separately, we hope that the recent addition of type-disambiguated array syntax opens the door to built-in support for uniform arrays (something we may end up implementing), and then these could come in both mutable and immutable variants. Similarly, it would be great to have immutable floatarrays, but this, too, is separate.

Reply:
I think that it is perfectly sound to have arrays of boxed floats, even in the default configuration -- this is what happens if you use Dynarrays for example, due to their use of magic to implement uniform arrays, similar to what is done in the Base library.

It is not *obvious* to me that covariance remains sound in presence of the dynamic float unboxing check -- non-uniform operations make everything harder to reason about. But I cannot see an issue as long as we assume (as is standard) that all types (involved in the subtyping relation) are "(float)-separated", that is, either they contain only floats or they have no float inhabitants. So I would guess that it is okay.

I'm convinced by the argument to keep things simple/separate by not mixing the two questions, so let me drop my suggestion to make immutable arrays non-unboxing.

Reply:
I'm glad you thought about the interaction between covariance and float-unboxing -- I had not! And now I have updated my internal definition of separability to include the subtype relation, which I had not realized (but easily agree with). In any case, based on your last comment, I think it's safe to Resolve this conversation.

Reply:
There's also an interaction between the recursive value check and float unboxing to consider: if immutable arrays support float unboxing then they'll be usable in fewer recursive contexts (which matters perhaps a smidgen more for immutable values.  See https://github.com/ocaml/ocaml/pull/1471#issuecomment-348814736 and https://github.com/ocaml/ocaml/pull/1516 for details.

Reply:
I would argue that it's better to not allow immutable structures to be defined recursively: for mutable ones, it's mostly convenience, as it's possible to get the same structures as long as you have a dummy value to put at the start. For immutable ones, it poses some tricky questions about the effects of the immutable reads (normally they have no effects or coeffects, but they must not be moved before the value gets fully initialised, and it's hard to express that properly in the backend). Plus, there are some functions that could be pure if immutable values were not allowed to be cyclic but cannot be considered so at the moment (for instance, `List.length` is not strictly pure; it wouldn't be a huge issue if we treated this one as pure but there are similar functions where this would break soundness).
So at least I would prefer if this PR did not allow more recursive definitions with immutable arrays than with regular arrays.